src/libexec/zmsearchndelete

```diff
@@ -0,0 +1,945 @@
+    warn( "$Prog: info: took ", timestr( timediff( $tEnd, $tStart ) ), "\n" );
+
+}
+
+################################################################################
+# subroutines
+
+# process_options will die via pod2usage on errors
+sub process_options {
+    my ( %Opt, @err );
+    GetOptions(
+        \%Opt,             "url=s",
+        "authuser=s",      "password=s",
+        "account=s@",      "exclude=s@",
+        "exclude-file=s@", "query=s",
+        "noadminauth",     "searchdirectory=s",
+        "help",            "man",
+        "debug:1",         "go",
+    ) or pod2usage( -verbose => 0 );
+
+    pod2usage( -message => "$Prog: version $VERSION\n", -verbose => 1 )
+      if ( $Opt{help} );
+    pod2usage( -verbose => 2 ) if ( $Opt{man} );
+
+    # set global DEBUG
+    $DEBUG = $Opt{debug} || 0;
+
+    foreach my $arg (qw(url authuser password query)) {
+        push( @err, "--$arg is required" ) if ( !$Opt{$arg} );
+    }
+
+    # special handling for --noadminauth and related account options
+    my $acctopts = "use of --account, --searchdirectory or <userlistfile>";
+    if ( $Opt{noadminauth} ) {
+        push( @err, "$acctopts is not supported with --noadminauth" )
+          if ( $Opt{account} or @ARGV or $Opt{searchdirectory} );
+
+        # for --noadminauth only allow/support searching that mailbox
+        $Opt{account} = [ $Opt{authuser} ] if $Opt{authuser};
+    }
+    else {
+        push( @err, "$acctopts is required" )
+          unless ( $Opt{account} or @ARGV or $Opt{searchdirectory} );
+    }
+
+    push( @err, "try --help option for more information" ) if (@err);
+
+    pod2usage(
+        -verbose => 0,
+        -message => join( "", map( "$Prog: $_\n", @err ) )
+    ) if (@err);
+
+    warn("$Prog: info: use '--go' option to actually DELETE matched messages\n")
+      unless ( $Opt{go} );
+
+    return %Opt;
+}
+
+# report search messages and delete it if --go is specified
+sub reportndelete {
+
+    my $opt = ( ref( $_[0] ) eq "HASH" ) ? shift : {};
+    my ( $rpt_data, $go ) = @_;
+
+    my ( $soap, $ainfo ) = @{$opt}{qw(soap ainfo)};
+    my ($noadminauth) = @{$opt}{qw(noadminauth)};
+    my $ahdr = $ainfo->{header};
+    my ( $dhdr, $msg_cnt );
+    warn(
+        "------------------------
+    1:Account\t2:MSG_ID\t3:CONV_ID\t4:DATE(mm/dd/yy)\t5:FROM\t6:SIZE(Bytes)\n"
+    );
+
+    #report searched message details
+    foreach my $acct ( sort keys(%$rpt_data) ) {
+        my $msg_ids;
+        my $msgs = $rpt_data->{$acct};
+        my $msgc = scalar(@$msgs);
+        $msg_cnt += $msgc;
+
+        warn("DEBUG: messages found in account: $acct : $msgc\n")
+          if ( $DEBUG > 0 );
+
+        if ($msgc) {
+            warn("------------------------\n");
+            foreach my $d (@$msgs) {
+
+                #my $tstr = scalar localtime( $d->{d} );
+                my $date = _date( $d->{d} );
+
+          #In testing, for very recently sent/received messages 'from' field
```

Comment:
nit: fix comment indentation to match code indentation (comment should start on the same 'column' as the code)
nit: also add a space after the '#' for all the comments in this part of the code (above and below)
example:
`   #some word => # some word`

Reply:
done